File: LeetCode/Python/Medium/287-Find-the-Duplicate-Number/solution.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        result: int=0

        seen: int=0

        for number in nums:
            #Get the bit in seen at the number-th position from the left
            bit_representing_number_in_seen: int=((seen >> number) & 1)

            #Check if the bit in seen at the number-th position from the left is 1
            already_seen_number: bool=(bit_representing_number_in_seen==1)

            #If so, we already seen the number and it is the duplicate
            if already_seen_number:
                result=number

                break
            #Else, set the bit in seen at the number-th position from the left to 1 to indicate that we've seen the number already
            else:
                seen=(seen | (1 << number))

        return result

File: LeetCode/Rust/Medium/287-Find-the-Duplicate-Number/solution.rs
impl Solution
{
    pub fn find_duplicate(nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let limit: usize=usize::pow(10, 5) + 1;

        let mut seen: Vec<bool>=vec![false; limit];

        for number in nums.iter()
        {   
            match seen[*number as usize]
            {
                true =>
                {
                    result=*number;

                    break;
                },
                false =>
                {
                    seen[*number as usize]=true;
                }
            };
        }

        result
    }
}

impl Solution
{
    pub fn find_duplicate(mut nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let n: usize=nums.len();

        for i in 0..n
        {
            let number: i32=nums[i].abs();
            
            let index: usize=(number - 1) as usize;

            nums[index]*=-1;

            if nums[index] > 0
            {
                result=number;

                break;
            }
        }

        result
    }
}